spring-integration-core/src/main/java/org/springframework/integration/history/MessageHistory.java

```diff
@@ -73,7 +78,26 @@ public static MessageHistory read(Message<?> message) {
 					new ArrayList<Properties>(previousHistory) : new ArrayList<Properties>();
 			components.add(metadata);
 			MessageHistory history = new MessageHistory(components);
-			message = messageBuilderFactory.fromMessage(message).setHeader(HEADER_NAME, history).build();
+
+			if (message instanceof MutableMessage) {
+				message.getHeaders().put(HEADER_NAME, history);
+			}
+			else if (message instanceof ErrorMessage) {
+				IntegrationMessageHeaderAccessor headerAccessor = new IntegrationMessageHeaderAccessor(message);
+				headerAccessor.setHeader(HEADER_NAME, history);
+				Throwable payload = ((ErrorMessage) message).getPayload();
+				ErrorMessage errorMessage = new ErrorMessage(payload, headerAccessor.toMessageHeaders());
+				message = (Message<T>) errorMessage;
+			}
+			else if (message instanceof AdviceMessage) {
+				IntegrationMessageHeaderAccessor headerAccessor = new IntegrationMessageHeaderAccessor(message);
+				headerAccessor.setHeader(HEADER_NAME, history);
+				message = new AdviceMessage<T>(message.getPayload(), headerAccessor.toMessageHeaders(),
+						((AdviceMessage) message).getInputMessage());
+			}
+			else {
+				message = messageBuilderFactory.fromMessage(message).setHeader(HEADER_NAME, history).build();
+			}
```

Comment:
I wonder if we should consider making the message history header mutable - via a `MessageHistoryHolder`; that way, we only have to rebuild the message in the first trackable component; each subsequent track would still get a new MessageHistory and each MB.build() would have to detect an existing holder, create a new one and transfer the existing history.

That way, pub/subs etc would inherit the history to-date, but have a unique holder so changes on one leg of the pub/sub would not be reflected in the other other leg. If the MB.build() didn't create a new holder, we'd get cross-talk.